avro-format.md

```diff
@@ -0,0 +1,176 @@
+# Avro Event Format for CloudEvents
+
+## Abstract
+
+The Avro Format for CloudEvents defines how events attributes are expressed in the [Avro 1.9.0 Specification][avro-spec].
+
+## Status of this document
+
+This document is a working draft.
+
+## Table of Contents
+
+1. [Introduction](#1-introduction)
+2. [Attributes](#2-attributes)
+3. [Examples](#3-examples)
+
+## 1. Introduction
+
+[CloudEvents][ce] is a standardized and transport-neutral definition of the
+structure and metadata description of events. This specification defines how the
+elements defined in the CloudEvents specification are to be represented in the
+[Avro 1.9.0][avro-primitives].
+
+The [Attributes](#2-attributes) section describes the naming conventions and
+data type mappings for CloudEvents attributes for use as Avro message
+properties.
+
+This specification does not define an envelope format. The Avro type system's
+intent is primarily to provide a consistent type system for Avro itself and not
+for message payloads.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+## 2. Attributes
+
+This section defines how CloudEvents attributes are mapped to the Avro
+type-system. This specification explicitly maps each attribute.
+
+### 2.1 Type System Mapping
+
+The CloudEvents type system MUST be mapped to Avro types as follows.
+
+| CloudEvents   | Avro                                                                   |
+| ------------- | ---------------------------------------------------------------------- |
+| Integer       | [int][avro-primitives]                                                 |
+| String        | [string][avro-primitives]                                              |
+| Binary        | [bytes][avro-primitives]                                               |
+| Map           | [map][avro-primitives]                                                 |
+| Any           | See [2.2](#22-mapping-any-typed-attributes)                            |
+| URI-reference | [string][avro-primitives] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [string][avro-primitives] following [RFC 3339][rfc3339] (ISO 8601)     |
+
+Extension specifications MAY define diverging mapping rules for the values of
+attributes they define.
+
+### 2.2 Mapping Any-typed Attributes
+
+`Any`-typed CloudEvents values can either hold a `String`, or a `Binary` value,
+or a `Map`. Avro type system SHOULD NOT satisfy the map-with-maps in a recursive
+way, i.e., every `map` declaration REQUIRE the `values` with primitives types
+or a declared type.
+
+### 2.3 Optional Attributes
+
+CloudEvents Spec defines optional attributes. The Avro format defines that
+every optional field MUST use the `null` type and the actual type through
+the [union][avro-unions].
```

Comment:
Does this mean that optional attributes will always appear, just with a "null" value? Is it a "MUST appear" or "MAY appear"? What happens if the send doesn't include it - are they violating something?

Reply:
@duglin it is "MAY appear". Using the `null` means that an optional attribute appears if it has value or no appear if there is no value at all.